File: packages/ts/src/hkdf.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { hkdf } from "@noble/hashes/hkdf.js";
import { sha256 } from "@noble/hashes/sha2.js";
import { ReallyMeCryptoError } from "./errors.js";

export const HKDF_MIN_INPUT_KEY_MATERIAL_LENGTH = 1;
export const HKDF_MAX_INPUT_LENGTH = 4096;
export const HKDF_MIN_OUTPUT_LENGTH = 1;
export const HKDF_MAX_OUTPUT_LENGTH = 4096;

export const ReallyMeHkdf = {
  deriveSha256(
    inputKeyMaterial: Uint8Array,
    salt: Uint8Array,
    info: Uint8Array,
    outputLength: number,
  ): Uint8Array {
    validate(inputKeyMaterial, salt, info, outputLength);
    return hkdf(sha256, inputKeyMaterial, salt, info, outputLength);
  },
} as const;

function validate(
  inputKeyMaterial: Uint8Array,
  salt: Uint8Array,
  info: Uint8Array,
  outputLength: number,
): void {
  if (
    inputKeyMaterial.length < HKDF_MIN_INPUT_KEY_MATERIAL_LENGTH ||
    inputKeyMaterial.length > HKDF_MAX_INPUT_LENGTH ||
    salt.length > HKDF_MAX_INPUT_LENGTH ||
    info.length > HKDF_MAX_INPUT_LENGTH ||
    !Number.isInteger(outputLength) ||
    outputLength < HKDF_MIN_OUTPUT_LENGTH ||
    outputLength > HKDF_MAX_OUTPUT_LENGTH
  ) {
    throw new ReallyMeCryptoError("invalid-input");
  }
}

File: packages/swift/Sources/ReallyMeCrypto/Hkdf.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import CryptoKit
import Foundation

public enum ReallyMeHkdf {
    public static let minInputKeyMaterialLength = 1
    public static let maxInputLength = 4096
    public static let minOutputLength = 1
    public static let maxOutputLength = 4096

    public static func deriveSha256(
        inputKeyMaterial: [UInt8],
        salt: [UInt8],
        info: [UInt8],
        outputLength: Int
    ) throws -> [UInt8] {
        try validate(
            inputKeyMaterial: inputKeyMaterial,
            salt: salt,
            info: info,
            outputLength: outputLength
        )
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: Data(inputKeyMaterial)),
            salt: Data(salt),
            info: Data(info),
            outputByteCount: outputLength
        )
        return key.withUnsafeBytes { bytes in
            Array(bytes)
        }
    }

    public static func deriveSha384(
        inputKeyMaterial: [UInt8],
        salt: [UInt8],
        info: [UInt8],
        outputLength: Int
    ) throws -> [UInt8] {
        try validate(
            inputKeyMaterial: inputKeyMaterial,
            salt: salt,
            info: info,
            outputLength: outputLength
        )
        let key = HKDF<SHA384>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: Data(inputKeyMaterial)),
            salt: Data(salt),
            info: Data(info),
            outputByteCount: outputLength
        )
        return key.withUnsafeBytes { bytes in
            Array(bytes)
        }
    }

    private static func validate(
        inputKeyMaterial: [UInt8],
        salt: [UInt8],
        info: [UInt8],
        outputLength: Int
    ) throws {
        guard (minInputKeyMaterialLength...maxInputLength).contains(inputKeyMaterial.count),
              salt.count <= maxInputLength,
              info.count <= maxInputLength,
              (minOutputLength...maxOutputLength).contains(outputLength)
        else {
            throw ReallyMeCryptoError.invalidInput
        }
    }
}
